syncer/error.go

```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
 	}
 	return err
 }
+
+// handleSpecialDDLError handles special errors for DDL execution.
+// it only ignore `invalid connection` error (timeout or other causes) for `ADD INDEX` now.
+// `invalid connection` means some data already sent to the server,
+// and we assume that the whole SQL statement has already sent to the server for this error.
+// if we have other methods to judge the DDL dispatched but timeout for executing, we can update this method.
+// NOTE: we must ensure other PK/UK exists for correctness.
+// NOTE: when we are refactoring the shard DDL algorithm, we also need to consider supporting non-blocking `ADD INDEX`.
+func (s *Syncer) handleSpecialDDLError(tctx *tcontext.Context, err error, ddls []string, index int, conn *DBConn) error {
+	// must ensure only the last statement executed failed with the `invalid connection` error
+	if len(ddls) == 0 || index != len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn {
+		return err // return the original error
+	}
+
+	parser2, err2 := s.fromDB.getParser(s.cfg.EnableANSIQuotes)
+	if err2 != nil {
+		return err // return the original error
+	}
+
+	ddl2 := ddls[index]
+	stmt, err2 := parser2.ParseOneStmt(ddl2, "", "")
+	if err2 != nil {
+		return err // return the original error
+	}
+
+	handle := func() {
+		tctx.L().Warn("ignore special error for DDL", zap.String("DDL", ddl2), log.ShortError(err))
+		err2 := conn.resetConn(tctx) // also reset the `invalid connection` for later use.
+		if err2 != nil {
+			tctx.L().Warn("reset connection failed", log.ShortError(err2))
+		}
+	}
+
+	switch v := stmt.(type) {
+	case *ast.AlterTableStmt:
+		// ddls should be split with only one spec
+		if len(v.Specs) > 1 {
+			return err
+		} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
+			handle()
+			return nil // ignore the error
+		}
+	case *ast.CreateIndexStmt:
+		handle()
+		return nil // ignore the error
+	default:
+		return err
```

Comment:
can remove the default case

Reply:
removed in https://github.com/pingcap/dm/pull/336/commits/81f4abb17f9fc03eed09a467aab6030ff1ffc805.